File: 0209-minimum-size-subarray-sum/0209-minimum-size-subarray-sum.cpp
class Solution {
public:
    int minSubArrayLen(int target, vector<int>& nums) {
        int n = nums.size(), ans = INT_MAX;
        vector<int> prefixSum(n + 1);
        for (int i = 0; i < n; ++i) prefixSum[i + 1] = prefixSum[i] + nums[i];
        
        for (int i = n; i >= 0 && prefixSum[i] >= target; --i){
            int left = 0, right = n;
            
            while (left < right){
                int mid = left + ((right - left) >> 1);
                if (prefixSum[i] - prefixSum[mid] >= target) left = mid + 1;
                else right = mid;
            }
            ans = min(ans, i - left + 1);
        }
        
        return ans == INT_MAX? 0 : ans;
    }
};

File: 0209-minimum-size-subarray-sum/0209-minimum-size-subarray-sum.py
class Solution:
    def minSubArrayLen(self, target: int, nums: List[int]) -> int:
        if sum(nums) < target: return 0
        
        ans, left, n = float("inf"), 0, len(nums)
        prefix_sum = [0] * (n+1)
        
        for i in range(1, n+1):
            prefix_sum[i] = prefix_sum[i-1] + nums[i-1]
        
        for right in range(1,n+1):
            while prefix_sum[right] - prefix_sum[left] >= target:
                left += 1
                ans = min(ans, right - left + 1)
        return ans